code/obj/submachine/slots.dm

```diff
@@ -6,155 +6,147 @@
 	var/plays = 0
 	var/working = 0
 	var/obj/item/card/id/scan = null
 
 	New()
 		AddComponent(/datum/component/mechanics_holder)
-		SEND_SIGNAL(src,COMSIG_MECHCOMP_ADD_INPUT,"activate", "activateinput")
 		..()
 
-	proc/activateinput(var/datum/mechanicsMessage/inp) //make this work some day.
-		//var/list/reflist = list("ops")
-		//Topic(null,
-		return
+/* INTERFACE */
 
-	attackby(var/obj/item/I as obj, user as mob)
-		if(istype(I, /obj/item/card/id))
-			if(src.scan)
-				boutput(user, "<span class='alert'>There is a card already in the slot machine.</span>")
-			else
-				boutput(user, "<span class='notice'>You insert your ID card.</span>")
-				usr.drop_item()
-				I.set_loc(src)
-				src.scan = I
-				src.updateUsrDialog()
-		else src.attack_hand(user)
+/obj/submachine/slot_machine/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if (!ui)
+		ui = new(user, src, "SlotMachine", name)
+		ui.open()
+
+/obj/submachine/slot_machine/ui_data(mob/user)
+	var/list/data = list()
+	data["busy"] = working
+	data["scannedCard"] = src.scan
+	data["money"] = src.scan?.money
+	data["plays"] = plays
+
+	return data
+
+/obj/submachine/slot_machine/ui_state(mob/user)
+	return tgui_physical_state
+
+/obj/submachine/slot_machine/ui_status(mob/user)
+  return min(
+		tgui_physical_state.can_use_topic(src, user),
+		tgui_not_incapacitated_state.can_use_topic(src, user)
+	)
+
+/obj/submachine/slot_machine/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	if(..())
 		return
+	switch(action)
+		if ("insert_card")
+			var/obj/O = usr.equipped()
+			if (istype(O, /obj/item/card/id))
+				boutput(usr, "<span class='notice'>You insert your ID card.</span>")
+				usr.drop_item()
+				O.set_loc(src)
+				src.scan = O
+				. = TRUE
+		if ("play")
+			if (src.working || !src.scan)
+				return TRUE
+			if (src.scan.money < 20)
+				src.visible_message("<span class='subtle'><b>[src]</b> says, 'Insufficient money to play!'</span>")
+				return TRUE
+			src.scan.money -= 20
+			src.plays++
+			src.working = 1
+			src.icon_state = "slots-on"
 
-	attack_hand(var/mob/user as mob)
-		src.add_dialog(user)
-		if (!src.scan)
-			var/dat = {"<B>Slot Machine</B><BR>
-			<HR><BR>
-			<B>Please insert card!</B><BR>"}
-			user.Browse(dat, "window=slotmachine;size=450x500")
-			onclose(user, "slotmachine")
-		else if (src.working)
-			var/dat = {"<B>Slot Machine</B><BR>
-			<HR><BR>
-			<B>Please wait!</B><BR>"}
-			user.Browse(dat, "window=slotmachine;size=450x500")
-			onclose(user, "slotmachine")
+			playsound(get_turf(src), "sound/machines/ding.ogg", 50, 1)
+			. = TRUE
+			ui_interact(usr, ui)
+			SPAWN_DBG(2.5 SECONDS) // why was this at ten seconds, christ
+				money_roll()
+				src.working = 0
+				src.icon_state = "slots-off"
+
+		if("eject")
+			if(!src.scan)
+				return TRUE // jerks doing that "hide in a chute to glitch auto-update windows out" exploit caused a wall of runtime errors
+			src.scan.set_loc(get_turf(src))
```

Comment:
I know this is parity with existing implementation, but would doing `usr.put_in_hand_or_eject(src.scan)` be better?